kerl

```diff
@@ -1292,52 +1228,119 @@ fi
 unset _KERL_VERSION _KERL
 
 set -o allexport
-_KERL_SAVED_REBAR_CACHE_DIR="\$REBAR_CACHE_DIR"
-_KERL_REBAR_CACHE_DIR_SET="\${REBAR_CACHE_DIR+x}"
-_KERL_SAVED_REBAR_CACHE_DIR="\$REBAR_CACHE_DIR"
-_KERL_REBAR_PLT_DIR_SET="\${REBAR_PLT_DIR+x}"
-_KERL_SAVED_REBAR_PLT_DIR="\$REBAR_PLT_DIR"
+
+_KERL_ACTIVE_DIR="$directory"
+
+_KERL_CLEANUP="unset _KERL_ACTIVE_DIR"
+
+if [ -n "\$BASH" ] || [ -n "\$ZSH_VERSION" ]; then
+    _KERL_CLEANUP="hash -r; \$_KERL_CLEANUP"
+fi
+
 _KERL_PATH_REMOVABLE="$directory/bin"
 PATH="\${_KERL_PATH_REMOVABLE}:\$PATH"
-_KERL_MANPATH_SET="\${MANPATH+x}"
-_KERL_MANPATH_REMOVABLE="$directory/lib/erlang/man:$directory/man"
-MANPATH="\${_KERL_MANPATH_REMOVABLE}:\$MANPATH"
-REBAR_CACHE_DIR="$directory/.cache/rebar3"
-REBAR_PLT_DIR="$directory"
-_KERL_ACTIVE_DIR="$directory"
+_KERL_CLEANUP="
+    PATH=\"\\\$(echo \"\\\$PATH\" | sed -e \"s%\$_KERL_PATH_REMOVABLE:%%\")\"
+    \$_KERL_CLEANUP
+"
+unset _KERL_PATH_REMOVABLE
+
 _KERL_ERL_CALL_REMOVABLE=\$(\\find $directory -type d -path "*erl_interface*/bin" 2>/dev/null)
 if [ -n "\$_KERL_ERL_CALL_REMOVABLE" ]; then
     PATH="\${_KERL_ERL_CALL_REMOVABLE}:\$PATH"
+    _KERL_CLEANUP="
+        PATH=\"\\\$(echo \"\\\$PATH\" | sed -e \"s%\$_KERL_ERL_CALL_REMOVABLE:%%\")\"
+        \$_KERL_CLEANUP
+    "
+fi
+unset _KERL_ERL_CALL_REMOVABLE
+
+_KERL_MANPATH_REMOVABLE="$directory/lib/erlang/man:$directory/man"
+if [ -n "\${MANPATH+x}" ]; then
+    if [ -n "\$MANPATH" ]; then
+        MANPATH="\${_KERL_MANPATH_REMOVABLE}:\$MANPATH"
+    else
+        MANPATH="\${_KERL_MANPATH_REMOVABLE}"
+    fi
+else
+    MANPATH="\${_KERL_MANPATH_REMOVABLE}"
+    _KERL_CLEANUP="
+        if [ -z \"\\\$MANPATH\" ]; then
+            unset MANPATH
+        fi
+        \$_KERL_CLEANUP
+    "
+fi
+_KERL_CLEANUP="
+    MANPATH=\"\\\$(echo \"\\\$MANPATH\" | sed -r \"s%\$_KERL_MANPATH_REMOVABLE:?%%\")\"
+    \$_KERL_CLEANUP
+"
+unset _KERL_MANPATH_REMOVABLE
+
+if [ -n "\${REBAR_PLT_DIR+x}" ]; then
+    _KERL_CLEANUP="REBAR_PLT_DIR=\"\$REBAR_PLT_DIR\"; \$_KERL_CLEANUP"
+else
+    _KERL_CLEANUP="unset REBAR_PLT_DIR; \$_KERL_CLEANUP"
+fi
+REBAR_PLT_DIR="$directory"
+
+if [ -n "\${REBAR_CACHE_DIR+x}" ]; then
+    _KERL_CLEANUP="REBAR_CACHE_DIR=\"\$REBAR_CACHE_DIR\"; \$_KERL_CLEANUP"
+else
+    _KERL_CLEANUP="unset REBAR_CACHE_DIR; \$_KERL_CLEANUP"
 fi
+REBAR_CACHE_DIR="$directory/.cache/rebar3"
+
 # https://twitter.com/mononcqc/status/877544929496629248
-_KERL_ERL_AFLAGS_SET="\${ERL_AFLAGS+x}"
 kernel_history=\$(echo "\$ERL_AFLAGS" | \\grep 'kernel shell_history' || true)
```

Comment:
For consistency, we should probably also rename this `_KERL_KERNEL_HISTORY`.

Reply:
@paulo-ferraz-oliveira I addressed most if not any of the concerns, the `add_cleanup` is really helpful, thank you.
Please let me know if anything else needs to be distilled.